Mlem/Views/Tabs/Search/SearchModel.swift

```diff
@@ -0,0 +1,111 @@
+    // constants
+    let pageSize: Int = 20
+    
+    // Used to switch tabs without refetching from API
+    var previousSearchText: String = ""
+    var firstPageCommunities: [AnyContentModel]?
+    var firstPageUsers: [AnyContentModel]?
+    
+    init(searchTab: SearchTab = .topResults) {
+        self.searchTab = searchTab
+    }
+    
+    func tabSwitchRefresh(contentTracker: ContentTracker<AnyContentModel>) {
+        switch searchTab {
+        case .topResults:
+            if let communities = firstPageCommunities, let users = firstPageUsers {
+                contentTracker.replaceAll(with: combineResults(communities: communities, users: users))
+                return
+            }
+        case .communities:
+            if let communities = firstPageCommunities {
+                contentTracker.replaceAll(with: communities)
+                return
+            }
+        case .users:
+            if let users = firstPageUsers {
+                contentTracker.replaceAll(with: users)
+                return
+            }
+        }
+        contentTracker.refresh(using: performSearch)
+    }
+    
+    func performSearch(page: Int) async throws -> [AnyContentModel] {
+        defer { previousSearchText = searchText }
+        switch self.searchTab {
+        case .topResults:
+            async let communities = try await searchCommunities(page: page)
+            async let users = try await searchUsers(page: page)
+            return try await combineResults(communities: communities, users: users)
+        case .communities:
+            if searchText != previousSearchText {
+                firstPageUsers = nil
+            }
+            return try await searchCommunities(page: page)
+        case .users:
+            if searchText != previousSearchText {
+                firstPageCommunities = nil
+            }
+            return try await searchUsers(page: page)
+        }
+    }
+    
+    @discardableResult
+    func searchCommunities(page: Int) async throws -> [AnyContentModel] {
+        let communities = try await apiClient.performSearch(
+            query: searchText,
+            searchType: .communities,
+            sortOption: .topAll,
+            listingType: .all,
+            page: page,
+            limit: pageSize
+        ).communities.map { AnyContentModel(CommunityModel(from: $0)) }
+        if page == 1 {
+            self.firstPageCommunities = communities
+        }
+        return communities
+    }
+    
+    @discardableResult
+    func searchUsers(page: Int) async throws -> [AnyContentModel] {
+        let users = try await apiClient.performSearch(
+            query: searchText,
+            searchType: .users,
+            sortOption: .topAll,
+            listingType: .all,
+            page: page,
+            limit: pageSize
+        ).users.map { AnyContentModel(UserModel(from: $0)) }
+        if page == 1 {
+            self.firstPageUsers = users
+        }
+        return users
+    }
```

Comment:
These calls should live in repositories that handle the API type -> internal model conversion